templates: change rawjson.tmpl to be backwards compatible

NDStrahilevitz opened the pull request:
Some users have reported a breaking change when we changed `Finding.Context` to `Finding.Event.Payload`.

This PR updates the `rawjson.tmpl` to be backwards compatible with to the previous struct(from 0.6.5). Later json templates need to be versioned appropriately (A discussion on this will be opened in the near future).

mtcherni95 reviewed (commented):
LGTM
@AsafEitani I suggest you to check it with your apps :)

NDStrahilevitz commented on cmd/tracee-rules/templates/0.6.5json.tmpl:
```diff
@@ -0,0 +1 @@
+{"Data": {{ toJson .Data }}\n"Context": {{ toJson .Event.Payload }}\n"SigMetadata": {{ toJson .SigMetadata }}\n}
```
My template was missing commas so it wasn't even a valid json 😅, thanks for noticing. I've pushed an updated template, have a look.

NDStrahilevitz replied:
Updated to what was in your initial comment.

danielpacak commented:
Just for my understanding. Is this PR related to the [rawjson.tmpl](https://github.com/aquasecurity/tracee/blob/v0.6.5/tracee-rules/templates/rawjson.tmpl) that will produce different output in the upcoming v0.7.0?

If yes, I'm wondering why don't we patch the rawjson.tmpl directly? Git revision will differentiate v0.6.5 and v0.7.0 versions so I'm not sure we need to append version numbers to template filenames:

https://github.com/aquasecurity/tracee/blob/v0.6.5/tracee-rules/templates/rawjson.tmpl
https://github.com/aquasecurity/tracee/blob/v0.7.0/tracee-rules/templates/rawjson.tmpl

NDStrahilevitz commented:
> Just for my understanding. Is this PR related to the [rawjson.tmpl](https://github.com/aquasecurity/tracee/blob/v0.6.5/tracee-rules/templates/rawjson.tmpl) that will produce different output in the upcoming v0.7.0?
> 
> If yes, I'm wondering why don't we patch the rawjson.tmpl directly? Git revision will differentiate between v0.6.5 and v0.7.0 so I'm not sure we need to append version numbers to template filenames:
> 
> https://github.com/aquasecurity/tracee/blob/v0.6.5/tracee-rules/templates/rawjson.tmpl https://github.com/aquasecurity/tracee/blob/v0.7.0/tracee-rules/templates/rawjson.tmpl

The usecase for this is for users of 0.7.0 that want their findings outputted in the format of 0.6.5 (their tools depend on the previous format). Since rawjson does a straight conversion to json, using that template from 0.6.5 on a 0.7.0 finding won't convert to the needed format.

danielpacak commented:
> > Just for my understanding. Is this PR related to the [rawjson.tmpl](https://github.com/aquasecurity/tracee/blob/v0.6.5/tracee-rules/templates/rawjson.tmpl) that will produce different output in the upcoming v0.7.0?
> > If yes, I'm wondering why don't we patch the rawjson.tmpl directly? Git revision will differentiate between v0.6.5 and v0.7.0 so I'm not sure we need to append version numbers to template filenames:
> > https://github.com/aquasecurity/tracee/blob/v0.6.5/tracee-rules/templates/rawjson.tmpl https://github.com/aquasecurity/tracee/blob/v0.7.0/tracee-rules/templates/rawjson.tmpl
> 
> The usecase for this is for users of 0.7.0 that want their findings outputted in the format of 0.6.5 (their tools depend on the previous format). Since rawjson does a straight conversion to json, using that template from 0.6.5 on a 0.7.0 finding won't convert to the needed format.
> 
> If someone wants to just output the current `Finding` struct to a json as is, then `rawjson.tmpl` will suffice, but if a previous version of the json is needed this new template will do that conversion.

Got it. If we want to maintain output templates for backward compatibility do you think it makes sense to unit test them? I saw in your previous comments that every single character matters ;)

NDStrahilevitz commented:
As discussed with @danielpacak, i've updated this PR to update `rawjson.tmpl` itself to be backwards compatible (by using @AsafEitani's tmpl code), and added a unit test to catch regressions in the future.

danielpacak commented on cmd/tracee-rules/output_test.go:
```diff
@@ -250,3 +250,58 @@ HostName: foobar.local
+		},
+	}
+
+	jsonTemplate, err := setupTemplate("templates/rawjson.tmpl")
+
+	if assert.NoError(t, err) {
+		for _, tc := range testCases {
+			//test if the raw json of findingv065 is equivalent to the 065 template result
+			var buf bytes.Buffer
+			jsonTemplate.Execute(&buf, tc.finding)
+
+			assert.JSONEq(t, tc.expectedJson, buf.String())
+		}
+	}
```
Usually we'd use require.NoError() to stop executing tests in case of critical failures. Also, for table driven tests using subtests with t.Run() will show you which one has actually failed in case of assertion errors:

```suggestion
	jsonTemplate, err := setupTemplate("templates/rawjson.tmpl")
	require.NoError(t, err)

	for _, tc := range testCases {
		t.Run(tc.testName, func(t *testing.T) {
			var buf bytes.Buffer
			err := jsonTemplate.Execute(&buf, tc.finding)
			require.NoError(t, err)
			assert.JSONEq(t, tc.expectedJson, buf.String())
		})
	}
```

NDStrahilevitz replied:
Right on both, i'll add it.

danielpacak commented on cmd/tracee-rules/output_test.go:
```diff
@@ -250,3 +250,58 @@ HostName: foobar.local
+					},
+				},
+			},
+			expectedJson: "{\"Data\":{\"a\":123,\"b\":\"c\",\"d\":true,\"f\":{\"123\":\"456\",\"foo\":\"bar\"}},\"Context\":{\"timestamp\":1321321,\"processorId\":0,\"processId\":21312,\"threadId\":0,\"parentProcessId\":0,\"hostProcessId\":0,\"hostThreadId\":0,\"hostParentProcessId\":0,\"userId\":0,\"mountNamespace\":0,\"pidNamespace\":0,\"processName\":\"\",\"hostName\":\"\",\"containerId\":\"abbc123\",\ ...
```
I believe this can be more readable with Go multiline strings, wdyt?

```suggestion
			expectedJson: `{
	"Data": {
		"a": 123
	}
}`,
```

NDStrahilevitz replied:
Done

itaysk commented:
came here to ask about tests and and saw this was already discussed. I agree it should have been fixed with this PR. @NDStrahilevitz are you working on it?

NDStrahilevitz commented:
> came here to ask about tests and and saw this was already discussed. I agree it should have been fixed with this PR. @NDStrahilevitz are you working on it?

Do you mean tests for the rest of the templates, or for the one modified in this PR? If the former, i'm not currently working on it (but I can open an issue on it), if the latter then a test was added as part of this PR.
